qchem/tests/test_deriv_grad.py

```diff
@@ -0,0 +1,96 @@
+import numpy as np
+import pytest
+
+import pennylane as qml
+from pennylane import qchem
+
+import shutil
+
+
+non_zero_coeffs = np.array(
+    [
+        -0.7763135743293005,
+        -0.08534360840293387,
+        -0.08534360840293387,
+        0.2669341092545041,
+        0.26693410925450134,
+        -0.025233628744274508,
+        0.0072162443961340415,
+        -0.0072162443961340415,
+        -0.0072162443961340415,
+        0.0072162443961340415,
+        -0.030654287745411964,
+        -0.023438043349280003,
+        -0.023438043349280003,
+        -0.030654287745411964,
+        -0.02494407786332001,
+    ]
+)
+non_zero_ops = [
+    qml.Identity(wires=[0]),
+    qml.PauliZ(wires=[0]),
+    qml.PauliZ(wires=[1]),
+    qml.PauliZ(wires=[2]),
+    qml.PauliZ(wires=[3]),
+    qml.PauliZ(wires=[0]) @ qml.PauliZ(wires=[1]),
+    qml.PauliY(wires=[0]) @ qml.PauliX(wires=[1]) @ qml.PauliX(wires=[2]) @ qml.PauliY(wires=[3]),
+    qml.PauliY(wires=[0]) @ qml.PauliY(wires=[1]) @ qml.PauliX(wires=[2]) @ qml.PauliX(wires=[3]),
+    qml.PauliX(wires=[0]) @ qml.PauliX(wires=[1]) @ qml.PauliY(wires=[2]) @ qml.PauliY(wires=[3]),
+    qml.PauliX(wires=[0]) @ qml.PauliY(wires=[1]) @ qml.PauliY(wires=[2]) @ qml.PauliX(wires=[3]),
+    qml.PauliZ(wires=[0]) @ qml.PauliZ(wires=[2]),
+    qml.PauliZ(wires=[0]) @ qml.PauliZ(wires=[3]),
+    qml.PauliZ(wires=[1]) @ qml.PauliZ(wires=[2]),
+    qml.PauliZ(wires=[1]) @ qml.PauliZ(wires=[3]),
+    qml.PauliZ(wires=[2]) @ qml.PauliZ(wires=[3]),
+]
+
+
+x = np.array([0.0, 0.0, 0.35, 0.0, 0.0, -0.35])
+
+
+def H(x):
+    return qchem.molecular_hamiltonian(["H", "H"], x)[0]
+
+
+@pytest.mark.parametrize(
+    ("i", "coeffs", "ops"),
+    [(0, [], []), (2, non_zero_coeffs, non_zero_ops), (5, -non_zero_coeffs, non_zero_ops)],
+)
+def test_deriv(i, coeffs, ops):
+    r"""Tests the correctness of the nuclear derivative of the electronic Hamiltonian
+    computed by the ``'derivative'`` function. 
+    """
+
+    deriv = qchem.derivative(H, x, i)
+
+    shutil.rmtree("pyscf")
+
+    exp_obs = qml.vqe.Hamiltonian(coeffs, ops)
+
+    assert exp_obs.compare(deriv)
```

Comment:
First time that I see use of `compare` in testing. We typically use `np.allclose`. Any reason for using `compare`?

Reply:
The ``compare`` method was brought to my attention by @antalszava and it is convenient to compare at once the ``coeffs`` and ``ops`` of two Hamiltonian objects. ``np.allclose`` is OK for the coefficients but not for the observables. This is going to change anyways @ixfoduap as we need to account for some tolerance for the coefficients.

Reply:
`compare` is being used for testing some `Hamiltonian` methods, see https://github.com/PennyLaneAI/pennylane/blob/c8673eb45d9b3d435867f04b9943a5c4b53322be/tests/test_vqe.py#L642

There might be better approaches though, even so, because as Alain mentioned, it's vulnerable to slight mismatches for the coefficients as it uses the `==` operator without any tolerance values and hence another approach using `np.allclose` is needed. Maybe adding another `compare_close` method or so that takes a tolerance value could be nice.